bigBed tileset service methods and tests

alexpreynolds opened the pull request:
## Description

_What was changed in this pull request?_

Added routines to generate tiles from pybbi-mediated interval queries to (particularly BED12-formatted) bigBed files. Added unit tests.

Tile queries are of the forms:

```
x.${z}.${y}
x.${z}.${y}.significant
x.${z}.${y}|max:A
x.${z}.${y}|min:A|max:B
x.${z}.${y}.significant|max:A
x.${z}.${y}.significant|min:A|max:B
```

The definitions of `z` and `y` represent tile zoom level and tile position.

The `significant` range key is the default, i.e. leaving it out does queries based on score significance stored in the bigBed score column (if present).

What is meant by "significant" is that, for the specified zoom level and tile position, a quantile is generated which returns the score that would threshold a maximum number of elements, either specified by `max` in the tile query, or by the constant `MAX_ELEMENTS` defined in `clodius/tiles/bigbed.py`.

If a tile query returns more than the selected maximum number of elements, then the set of elements is thresholded by the quantile-derived score, and a uniformly random sample is drawn from the set of elements, after the score threshold is applied, to ensure that the set is no larger than the requested maximum number of elements.

_Why is it necessary?_

To begin to add support for bigBed queries by higlass-server, and ultimately gene annotation track visualizations in the higlass plugin.

## Checklist

- [x] Unit tests added or updated
- [x] Updated CHANGELOG.md

pkerpedjiev commented on clodius/tiles/bigbed.py:
```diff
@@ -0,0 +1,407 @@
+logger = logging.getLogger(__name__)
+
+range_modes = {}
+range_modes['significant'] = {'name': 'Significant', 'value': 'significant'}
```
The PR description says `significance` and one of the tile queries below also uses `significance`. Is it intentional to use `significant` here?

alexpreynolds replied:
I think I have a typo in my PR description. I use `significant` in the module and in the associated test.

pkerpedjiev commented on clodius/tiles/bigbed.py:
```diff
@@ -0,0 +1,407 @@
+range_modes['significant'] = {'name': 'Significant', 'value': 'significant'}
+
+
+def get_quadtree_depth(chromsizes):
```
This function and a few of the others look like they're identical to the ones in `clodius.tiles.bigWig`. Maybe use those directly with an import or factor out to a common file?

alexpreynolds replied:
I'll import from `clodius.tiles.bigWig`.

pkerpedjiev commented on clodius/tiles/bigbed.py:
```diff
@@ -0,0 +1,407 @@
+        
+        fetch_factory = ft.partial(bbi.fetch_intervals, bbpath, chrom, start, end)
+
+        if range_mode == 'significant':
```
`significant` vs. `significance`?

pkerpedjiev replied:
And since there's no difference between including that option and not including it, maybe it's better to leave it out until it causes different behavior?

alexpreynolds replied:
Unless you have a strong objection to this, I suggest leaving in a descriptive name for the range mode, as there may be other approaches used for querying and subsetting ranges that have not been considered, and this helps document what the current approach is (even if it is the only option I wrote, and the default one).

nvictus replied:
I agree that keeping it makes sense. Could you document how the `significance` filter is downsampling the intervals in a tile? If it's not obvious where to put it, you could put it in a module-level docstring for now.

I also don't know if bigBeds are guaranteed to have a column 5 or that it be an integer `score` column. Maybe there needs to be a fallback to treating such files as having score = 1 for uniform downsampling.

pkerpedjiev commented on clodius/tiles/bigbed.py:
```diff
@@ -0,0 +1,407 @@
+        for interval in intervals2:
+            final_intervals.append({'chrOffset': chrOffsets[chrom], 'importance': int(interval[4]), 'fields': interval})
+            
+    elif intervals_length > max_elements:
```
Comments in the section below would be helpful for understanding what's happening.

pkerpedjiev replied:
Actually, in looking over it, it's actually pretty clear what's happening. Feel free to ignore the above comment :-)

pkerpedjiev commented on clodius/tiles/bigbed.py:
```diff
@@ -0,0 +1,407 @@
+        thresholded_intervals = []
+        desired_perc = max_elements / intervals_length
+        thresholded_score = int(np.quantile(scores, 1-desired_perc))
+        for interval in intervals2:
```
Might be clearer to use a list comprehension here:

```
thresholded_intervals = [interval for interval in intervals2 if int(interval[4]) >= thresholded_score]
```

alexpreynolds replied:
I used a list comprehension, slightly modified to return `beddb`-like elements.

alexpreynolds replied:
I do not know of a way to do try/except handling within a list comprehension, so in case the bigBed data lacks score data, or score data are not integer-like, I reverted back to a `for` loop in order to handle that case.

pkerpedjiev commented on clodius/tiles/bigbed.py:
```diff
@@ -0,0 +1,407 @@
+            if int(interval[4]) >= thresholded_score:
+                thresholded_intervals.append({'chrOffset': chrOffsets[chrom], 'importance': int(interval[4]), 'fields': interval})
+        thresholded_intervals_length = len(thresholded_intervals)
+        if thresholded_intervals_length > max_elements:
```
I guess this is for the case when you have multiple equal scores so that your threshold doesn't draw a clear line in the number above and below?

alexpreynolds replied:
Yes, the score threshold can easily get more than the maximum number of elements, especially when zoomed out.

pkerpedjiev commented on clodius/tiles/bigbed.py:
```diff
@@ -0,0 +1,407 @@
+    return generated_tiles
+
+
+def chromsizes(filename):
```
Just reroute to the bigWig function, no?

alexpreynolds replied:
I added another commit to call the bigwig code for `get_chromsizes`.

alexpreynolds commented:
I ended up reverting most of these commits as they broke serving elements through higlass-server. Still investigating.

alexpreynolds commented:
Possible code style issues?

```
flake8
./clodius/tiles/bigbed.py:6:1: F401 're' imported but unused
./clodius/tiles/bigbed.py:7:1: F401 'sys' imported but unused
./clodius/tiles/bigbed.py:41:5: F841 local variable 'n_bins' is assigned to but never used
./clodius/tiles/bigbed.py:45:9: F841 local variable 'clen' is assigned to but never used
./clodius/tiles/bigbed.py:46:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:60:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:64:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:72:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:76:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:79:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:83:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:88:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:92:54: E226 missing whitespace around arithmetic operator
./clodius/tiles/bigbed.py:98:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:114:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:124:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:140:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:168:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:171:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:188:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:192:45: W291 trailing whitespace
./clodius/tiles/bigbed.py:196:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:226:1: W293 blank line contains whitespace
./clodius/tiles/bigbed.py:239:1: W293 blank line contains whitespace
./test/tiles/bigbed_test.py:3:1: F401 'numpy as np' imported but unused
./test/tiles/bigbed_test.py:4:1: F401 'base64' imported but unused
./test/tiles/bigbed_test.py:5:1: F401 'sys' imported but unused
./test/tiles/bigbed_test.py:7:1: E302 expected 2 blank lines, found 1
./test/tiles/bigbed_test.py:12:1: W293 blank line contains whitespace
./test/tiles/bigbed_test.py:19:1: W293 blank line contains whitespace
./test/tiles/bigbed_test.py:31:33: E231 missing whitespace after ','
./test/tiles/bigbed_test.py:31:40: E231 missing whitespace after ','
The command "./travis_test.sh" exited with 1.
```

pkerpedjiev commented on clodius/tiles/bigbed.py:
```diff
@@ -0,0 +1,239 @@
+    range_mode=None,
+    min_elements=None,
+    max_elements=None
+):
```
I know we don't have a strong tradition of doing this right now, but perhaps you could add a docstring here describing what the return type looks like?

pkerpedjiev commented:
Looks much better 👍 Thanks for making all those changes.

I just added one more comment about documenting the return type of the tile function. It looks like you're returning an array for each chromosome which differs from what we do in the traditional gene annotations track where we just return one array containing all the entries (i.e. not broken down by chromosome).

alexpreynolds commented:
I added an explanatory docstring and added exception handling for scoreless or non-integer score data in the source bigBed. 

I do appear to be returning an array for each chromosome. I am not intentionally doing this, but this appears to be a consequence of how querying is done here on queries spanning chromosomes, and that starts and ends are provided for each chromosome:

https://github.com/alexpreynolds/clodius/blob/a87fb82077a11c19cb557d7ed3a62a9d6c9ebbf8/clodius/tiles/bigbed.py#L177-L191

I had tried collapsing this list-of-lists to one list before passing it to `higlass-server`, but there are problems on the `higlass-server` side, where I have difficulties getting Redis to correctly deserialize what it caches.

A collapsed list which is `get` from Redis looks like it is split into separate characters:

https://github.com/alexpreynolds/higlass-server/blob/bigbed-support/tilesets/views.py#L437-L438

If I `set` a list-of-lists as they are, then that can be correctly `get` via Redis later on in the block of code linked above.

Here is where the data are `set` if not previously cached: 

https://github.com/alexpreynolds/higlass-server/blob/bigbed-support/tilesets/views.py#L479-L487

When later retrieved, either through a direct query from the bigBed file, or deserialized from Redis, I do indeed collapse after the fact:

https://github.com/alexpreynolds/higlass-server/blob/bigbed-support/tilesets/views.py#L489-L491

This collapsed list does look like what the current beddb gene annotation track parses, which is why I collapse it here.

I'm happy to do this a different and less convoluted way, but I would ask for advice in how to collapse the list-of-lists on the clodius side, in such a way that Redis can serialize and deserialize it properly?

pkerpedjiev commented:
Would you be able to post the error that redis throws? It's strange that it can deserialize a list of lists but not a list.

Have you had a chance to look at how `clodius.tiles.beddb` formats its return values? https://github.com/higlass/clodius/blob/268f4adc5e662b3e2779d6b8a07f59a8e0b79325/clodius/tiles/beddb.py#L74

IMO, if possible, it would be better to do the collapsing in `clodius` so that higlass-server just responds to request without further formatting returned tiles.

alexpreynolds commented:
Wouter Meuleman has asked that I examine an aggregation method that is applied at higher zoom levels (levels closer to whole-genome scale) — perhaps something that counts intervals over bins and which the higlass BED12 plug-in (PR forthcoming) would render as a 1D density-like heatmap or line plot. 

I'm not sure if you'd prefer leaving this PR open as I work on this, or if you'd like this PR to be wrapped in sooner? 

In my opinion, having this PR wrapped in would make it easier to get the higlass-server PR worked in (and to submit a PR for the higlass visualization piece, itself). 

I can then add the aggregation feature after, as the next stage of bigBed support.

(I guess I'm asking if you folks are generally satisfied with this PR as it is now?)

pkerpedjiev commented:
Sorry, completely lost track of this PR. I just merged it and will cut a new release once I finish replying here :-) As always, please feel free to send reminders if I don't respond to something within a day or two.

For the other aggregation method, I was playing around with something even simpler than displaying interval density. What it does is merge intervals that are closer than some number of pixels apart visually. It's simple in concept and best demonstrated with an example:

https://observablehq.com/d/cee1a89063441de0

The logic is that the vast majority of the time you only care if there is *a* feature somewhere and not what the density of intervals is. This aggregation method introduces new features of type "filler" which just denote regions with features in them. So if you're zoomed way out and some features are hidden, the filler region that contains them will still be shown indicating that if you zoom in, you'll find more data.

This is already implemented for the Genbank data fetcher in higlass here: https://github.com/higlass/higlass/blob/49e52d1c77ca7fbd2a74b711f36ccbc24c7a46b3/app/scripts/data-fetchers/genbank-fetcher.js#L14

There is also a PR open to update the Gene Annotations track to support this datatype but I haven't merged it yet: https://github.com/higlass/higlass/pull/735

Take a look and let me know what you think.
